rocAL/source/decoders/video/rocdec_video_decoder.cpp

```diff
@@ -0,0 +1,184 @@
+    _demuxer = std::make_shared<VideoDemuxer>(src_filename);
+    rocDecVideoCodec rocdec_codec_id = AVCodec2RocDecVideoCodec(_demuxer->GetCodecID());
+    _rocvid_decoder = std::make_shared<RocVideoDecoder>(device_id, mem_type, rocdec_codec_id, 0, nullptr, 0);
+
+    if(!_rocvid_decoder->CodecSupported(device_id, rocdec_codec_id, _demuxer->GetBitDepth())) {
+        WRN("GPU doesn't support codec!")
+        return VideoDecoder::Status::FAILED;
+    }
+
+    std::string device_name, gcn_arch_name;
+    int pci_bus_id, pci_domain_id, pci_device_id;
+    _rocvid_decoder->GetDeviceinfo(device_name, gcn_arch_name, pci_bus_id, pci_domain_id, pci_device_id);
+    std::cout << "info: Using GPU device " << device_id << " - " << device_name << "[" << gcn_arch_name << "] on PCI bus " <<
+    std::setfill('0') << std::setw(2) << std::right << std::hex << pci_bus_id << ":" << std::setfill('0') << std::setw(2) <<
+    std::right << std::hex << pci_domain_id << "." << pci_device_id << std::dec << std::endl;
+
+    return status;
+}
+
+// Seeks to the frame_number in the video file and decodes each frame in the sequence.
+VideoDecoder::Status RocDecVideoDecoder::Decode(unsigned char *out_buffer, unsigned seek_frame_number, size_t sequence_length, size_t stride, int out_width, int out_height, int out_stride, AVPixelFormat out_pix_format) {
+    
+    VideoDecoder::Status status = Status::OK;
+    VideoSeekContext video_seek_ctx;
+
+    // Reconfig the decoder
+    ReconfigDumpFileStruct reconfig_user_struct = { 0 };
+    ReconfigParams reconfig_params = { 0 };
+    reconfig_params.p_fn_reconfigure_flush = ReconfigureFlushCallback;
+    reconfig_params.reconfig_flush_mode = RECONFIG_FLUSH_MODE_NONE;
+    reconfig_params.p_reconfig_user_struct = &reconfig_user_struct;
+    _rocvid_decoder->SetReconfigParams(&reconfig_params);
+
+    if (!_demuxer || !_rocvid_decoder || !out_buffer) {
+        ERR("Decoder is not initialized");
+        return Status::FAILED;        
+    }
+    if (!out_buffer || !(sequence_length|stride)) {
+        ERR("Invalid parameter passed");
+        return Status::FAILED;        
+    }
+
+    int64_t pts = 0, dts = 0, requested_frame_dts = 0;
+    int n_video_bytes = 0, n_frame_returned = 0, n_frame = 0, pkg_flags = 0;
+    OutputSurfaceInfo *surf_info;
+    uint8_t *pvideo = nullptr;
+    int num_decoded_frames = sequence_length * stride;
+
+    bool b_seek = true;       // seek only for first frame in sequence
+    uint32_t image_size = out_height * out_stride * sizeof(uint8_t);
+    video_seek_ctx.seek_crit_ = SEEK_CRITERIA_FRAME_NUM;
+    video_seek_ctx.seek_mode_ = SEEK_MODE_PREV_KEY_FRAME;
+    VideoPostProcess post_process;
+    bool sequence_decoded = false;
+    do {
+        if (b_seek) {
+            video_seek_ctx.seek_frame_ = static_cast<uint64_t>(seek_frame_number);
+            video_seek_ctx.seek_crit_ = SEEK_CRITERIA_FRAME_NUM;
```

Comment:
Do we always seek with frame number? Is there a use case for seek with timestamp?

Reply:
In rocAL we generally split the sequences based on the frame number and seek wrt frame number.
Seeking with timestamp is not used here